README.md

```diff
@@ -12,15 +12,18 @@ go-buffer-pool
 
 ## Table of Contents
 
-- [Use Case](#use-case)
+- [About](#about)
     - [Advantages over GC](#advantages-over-gc)
     - [Disadvantages over GC:](#disadvantages-over-gc)
 - [Contribute](#contribute)
 - [License](#license)
 
-## Use Case
+## About
 
-Use this when you need to repeatedly allocate and free a bunch of temporary buffers of approximately the same size.
+This library provides:
+
+1. A buffer pool for re-using byte buffers of _varying_ sizes. Unlike `sync.Pool`, this pool will always return an appropriately sized buffer (the next power of two) so it's appropriate for applications that need to pool dynamically sized buffers.
+2. A `bytes.Buffer` implementation backed by a buffer pool. Unlike `bytes.Buffer`, `pool.Buffer` will automatically "shrink" on read, using the buffer pool to avoid causing too much work for the allocator. This is primarily useful for long lived buffers that usually sit empty.
```

Comment:
I don't think it's entirely correct to describe anything as "a `bytes.Buffer` implementation" as `bytes.Buffer` is a concrete type. Maybe: "A type similar to `bytes.Buffer` ..."?

I also don't really understand the use case for this, either :) Just out of curiosity, do you have a representative example handy?

Reply:
> I don't think it's entirely correct to describe anything as "a bytes.Buffer implementation" as bytes.Buffer is a concrete type. Maybe: "A type similar to bytes.Buffer ..."?

Eh, I think people can figure it out but I can tweak it a bit.

> I also don't really understand the use case for this, either :) Just out of curiosity, do you have a representative example handy?

We used to use it in one of our stream multiplexers (we don't now because we needed something fancier).

1. With a normal buffer, the size of the buffer will grow to the size of the largest "burst" and never shrink.
2. With these buffers, the size will shrink back down.

Reply:
Does shrinking a buffer not invoke GC?

Reply:
We put the buffer back into a pool. If nobody needs it, it gets garbage collected. If we're allocating a lot of buffers, it'll get reused.

Reply:
So my understanding is that a pool.Buffer manages a set of power-of-2-sized buffers, each backed by a sync.Pool, and gives you an appropriately-sized one of those based on the size you request. Is that accurate? If so, I don't quite get what "a pool.Buffer will automatically 'shrink' on read" means, nor how this represents less work for the allocator. It seems more or less equivalent to an explicit set of size-classed sync.Pools. But I'm sure I'm missing something!

Reply:
A `pool.Buffer` has a reference to a `pool.BufferPool` which is the set of sync.Pools for different size byte slices.

A `pool.Buffer` will grow or shrink, not by simply allocating a bigger byte slice or creating a new smaller byte slice, but by using the referenced `pool.BufferPool` to get a byte slice. If you can re-use a byte slice from `pool.BufferPool` that's one less alloc you need to pay.

Of course you can reimplement this with a set of size-classed sync.Pools, this is what this library is. You just need to be careful not to allocate accidentally and take care when growing/shrinking.

The code is pretty simple, and I recommend taking a look and tweaking it. The tests cases are pretty nice so you'll see how things break (nice work @Stebalien).